Add Progressive Web App (PWA) functionality

caleb531 opened the pull request:
- Add a web app manifest
- Add service worker for offline access

With these two features in place, Yugioh Calculator will be installable as a native app on Chrome and Android. The installed app will also utilize the icon and defined colors to present a splash screen while loading.

Yugioh Calculator now scores a 93/100 on the PWA reporting tool, Lighthouse. The JSON for the above report is attached below. To view it, drag it into the [online viewer](https://googlechrome.github.io/lighthouse/viewer/).

[localhost_2017-05-15_20-21-41.json.zip](https://github.com/jacksonrayhamilton/yugioh-calculator-2016/files/1003124/localhost_2017-05-15_20-21-41.json.zip)

<img width="736" alt="screen shot 2017-05-15 at 8 30 04 pm" src="https://cloud.githubusercontent.com/assets/872474/26088659/4ec60d86-39ad-11e7-9c6f-b81178f747fb.png">

Currently, the service worker does not handle Google Analytics. However, there *is* [such a package](https://developers.google.com/web/updates/2016/07/offline-google-analytics) for offline GA. I will explore this option soon.

caleb531 committed:
Display friendly message when JS is disabled

caleb531 committed:
Add web app manifest

The manifest enables Yugioh Calculator to be installable on supported
browsers. The supplied colors and icon allow the installed "native" app
to display a custom splash screen upon startup.

caleb531 committed:
Add service worker for offline access

jacksonrayhamilton commented on Gruntfile.js:
```diff
@@ -21,7 +21,7 @@ module.exports = function (grunt) {
           html: 'public/index.html'
         },
-        src: 'design/icon.png',
+        src: 'app/assets/web-app-icon.png',
```
This breaks favicon generation.  `icon.16x16.png` is no longer used for the generated `favicon.ico`.

That said, comparing `icon.16x16.png` and the generated imagemagick icon, I think the quality of the latter is good enough.  So can you delete `icon.16x16.png` please?

caleb531 replied:
Done! :)

jacksonrayhamilton commented on app/app.js:
```diff
@@ -195,6 +195,10 @@ function App (spec) {
     document.removeEventListener('keydown', onKeydown);
     m.mount(element, null);
   };
+  // Register service worker to enable offline access
```
End explanatory fragments with a period, please.

jacksonrayhamilton commented on app/assets/index.html:
```diff
@@ -9,9 +9,11 @@
     <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent"/>
     <meta name="apple-mobile-web-app-title" content="Yugioh Calc"/>
     <meta name="format-detection" content="telephone=no"/>
+    <meta name="theme-color" content="#ffffff" />
```
Inconsistent space before `/>`.

jacksonrayhamilton commented on app/assets/index.html:
```diff
@@ -22,6 +24,9 @@
   </head>
   <body>
+    <noscript>
+      <p style="text-align: center;">Please enable JavaScript to use Yugioh Calculator.</p>
```
Can you use a CSS class and add more styling here?

A "reload" button might also be nice (i.e. an HTML link pointing to the same page).

I am thinking that a light gray rectangle with rounded corners (like the one used for life points) would be in the top part of the screen instructing to enable JS, and a dark gray rectangle with rounded corners (like the one used for digits) would be in the lower part of the screen and provide the "reload" button.

[General idea.](https://shareitdammit.com/MIoJiIfm)  Feel free to improve or be more creative.

caleb531 replied:
A "reload" button won't work if JS is disabled, but yes, I'll add prettier and more-consistent styling.

jacksonrayhamilton replied:
This wouldn't work with JS disabled?

```html
<a class="button" href="/">Reload</a>
```

jacksonrayhamilton reviewed (changes requested):
A few things that I'd like addressed but do feel free to oppose me.

Also I'll wait until the GA thing to merge.

caleb531 commented on Gruntfile.js:
```diff
@@ -21,7 +21,7 @@ module.exports = function (grunt) {
           html: 'public/index.html'
         },
-        src: 'design/icon.png',
+        src: 'app/assets/web-app-icon.png',
```
Darn, I forgot to update the icon path in `Gruntfile.js`—do you still want me to do that, or is it irrelevant now that you are declaring `icon-16x16.png` obsolete?

caleb531 committed:
End service worker comment with period

caleb531 committed:
Style theme color self-closing tag consistently

caleb531 committed:
Consolidate and improve noscript message styling

caleb531 committed:
Delete 16x16 icon, as requested

jacksonrayhamilton commented on app/styles/theme.css:
```diff
@@ -33,6 +33,20 @@ body {
+  background-color: #eeeeee;    /* MD Grey 200*/
+  font-size: 5vh;
+  text-align: center;
+}
```
The purpose of `theme.css` is to apply styling like colors and spacing all in one place, and this centralization is supposed to help keep that styling consistent.  Part of this is due to the decision to use plain CSS, which is half tied up with the project's history, and half as an experiment.

Anyway, properties declared here that you see declared elsewhere in this file should stay here, and probably the rest should be moved to a file called `noscript.css`.

You may also notice that for essentially all other styles where it was possible, I opted to use classes instead of element selectors.  I believe this is a more reliable way of writing CSS for a web app, and would prefer if these styles could be declared under class selectors instead.

caleb531 committed:
Moved noscript styling; added reload button

caleb531 committed:
Merge branch 'master' into pwa

jacksonrayhamilton commented:
Two issues with the noscript styling:

1) `<div class="yc-app"></div>` is still on the page, which doubles the vertical space and creates scrollbars.

2) I'd prefer if the rectangles were the same width.

caleb531 committed:
Move noscript inside of div.yc-app

Previously, when JS was disabled, the fact that div.yc-app was next to
the noscript container caused the page to overflow. Moving the noscript
inside of this container fixes the issue.

caleb531 commented:
Done! Also, feel free to merge this PR whenever you're ready (or perhaps once you've performed your extensive benchmarking!).

caleb531 commented:
And... I forgot to address your point about the rectangles being of equal width. Stand by for a commit to fix that...

caleb531 committed:
Make noscript message and button the same width

caleb531 commented:
Okay, *now* it is finished! Enjoy your new Progressive Yugioh Calculator! :)

jacksonrayhamilton commented:
The styling leaves a bit to be desired on what Chrome claims to be the iPad screen resolution: https://shareitdammit.com/qio77uFO  Will you ensure it looks pleasing on all common screen resolutions?

jacksonrayhamilton commented on app/styles/noscript.css:
```diff
@@ -0,0 +1,23 @@
+.no-js-container {
```
All other classes in this app are prefixed with "yc-", so these classes should start with "yc-no-js-", instead.

jacksonrayhamilton commented on app/styles/noscript.css:
```diff
@@ -0,0 +1,23 @@
+  margin: 10px auto 10px auto;
+  border-radius: 3px;
+  padding: 20px;
+  background-color: #eeeeee;    /* MD Grey 200*/
```
Missing space before closing `*/`.

jacksonrayhamilton reviewed (changes requested):
More trivial updates.

jacksonrayhamilton committed:
Add a script and instructions for testing HTTPS-only features.

jacksonrayhamilton commented:
I was able to test this with https://github.com/caleb531/yugioh-calculator-2016/pull/3.  Indeed, the app now works offline on Android, both in Android Chrome and when added to the home screen (which I was immediately prompted to do at https://jacksonrayhamilton.github.io/yugioh-calculator-2016/).

Thanks for your work on this so far!

caleb531 committed:
Merge pull request #3 from jacksonrayhamilton/pwa

Add a script and instructions for testing HTTPS-only features.

caleb531 committed:
Add yc- prefix to noscript class names

caleb531 committed:
Fix noscript text overflow on larger displays

caleb531 committed:
Add missing spaces before */

caleb531 commented:
Alright, missing spaces added, classes prefixed, and noscript text overflow fixed:

![screen shot 2017-06-10 at 11 13 47 am](https://user-images.githubusercontent.com/872474/27005184-f45b879c-4dcd-11e7-8ca8-444ba3d18d43.png)

![screen shot 2017-06-10 at 11 13 40 am](https://user-images.githubusercontent.com/872474/27005185-f65a8c0a-4dcd-11e7-8941-87c1fc5aa32a.png)

jacksonrayhamilton commented:
It is done.
